src/unknown.rs

```diff
@@ -0,0 +1,149 @@
+use alloc::vec::Vec;
+
+use core::fmt::Debug;
+use core::usize;
+
+use bytes::{Buf, BufMut};
+
+use crate::encoding::{decode_varint, encode_key, encode_varint, key_len, WireType};
+use crate::DecodeError;
+
+pub struct UnknownField {
+    tag: u32,
+    wire_type: WireType,
+    bytes: Vec<u8>,
+}
+
+impl Debug for UnknownField {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("UnknownField")
+            .field("tag", &self.tag)
+            .field("wire_type", &self.wire_type)
+            .field("bytes", &self.bytes)
+            .finish()
+    }
+}
+
+impl Clone for UnknownField {
+    fn clone(&self) -> Self {
+        Self {
+            tag: self.tag.clone(),
+            wire_type: self.wire_type.clone(),
+            bytes: self.bytes.clone(),
+        }
+    }
+}
+
+impl PartialEq for UnknownField {
+    fn eq(&self, other: &Self) -> bool {
+        self.tag == other.tag && self.wire_type == other.wire_type && self.bytes == other.bytes
```

Comment:
i dont actually think this is correct. without knowing the message each field originated from, the fact that they have the same tag, wire type, and bytes doesnt mean they _are_ the same field semantically. you can make the choice that this is how you implement it, but choosing that unknown things are not Eq at all is equally valid.

Reply:
I see your point, and I see why electing not to implement `PartialEq` could be a reasonable choice.  For example, if I received a `Foo` message and a `Bar` message, each with a single unknown field at tag `2` with wire type `LEN` and zero bytes, I wouldn't know if these fields represent the same thing.  Indeed, maybe `Foo` includes an empty string, but `Bar` includes a default value `Baz` message.  It might be safest to just not compare these unknown fields.

On the other hand, it might be reasonable to say that _qua_ protobuf unknown field, these really are the same thing.  An analogy: the semantics of the number `42` are different in different places (e.g. in one struct it might represent a sequential id, but in another it might represent a quantity), but _qua_ number, `42` is `42`.  Of course, I recognize that there are deficiencies in this analogy, and it is hardly dispositive.

That said, one of the reasons I would like to implement `PartialEq` for unknown fields is to avoid breaking the `PartialEq` implementations of structs in generated code.  If I update my `build.rs` script so that my generated structs include unknown fields, I don't want to lose the ability to compare generated structs with each other.  I see at least a couple of options here:

1. Give up the `PartialEq` impl for these generated structs.
2. Generate a custom `PartialEq` impl for the generated struct which does not compare the unknown fields (ie. two structs which vary only in the unknown fields would be considered equal).
3. Generate a custom `PartialEq` impl for the generated struct which compares the unknown fields as if there were a `PartialEq` impl for them (compare tag, wiretype, and bytes), but don't actually provide a `PartialEq` impl for the `UnknownField` and `UnknownFields` structs.
4. Include a `PartialEq` impl for `UnknownFields` and `UnknownField`.

There may also be some other options which involve generating message-specific unknown fields types (e.g. `FooUnknownFields` for the `Foo` struct), or other approaches.

The present PR suggests (4).  You have noted that (1) is also a reasonable choice.  Between (2) and (3), I would favor (3), since I can easily imagine being surprised to find that `Foo` with some unknown fields is the "same" as `Foo` without any unknown fields.

The present PR does not attempt to fully deserialize the unknown fields. That is, if an unknown field contains a message, the present PR would just create a single unknown field struct for that nested message, no matter how big and deep it is; indeed, I don't think we _could_ fully deserialze things, at the risk of mistaking a string for a message (though I could be mistaken about this; using the `--decode-raw` option with `protoc` seems to be able to distinguish the two).  Because of this, there are some other deficiencies in (3) and (4).  In particular, if an unknown field contains a message, the present implementation would consider two otherwise identical messages to be different if their field order over the wire differed.

Reply:
i actually would favour (2), for whats its worth. i feel that Eq on the _message_ containing unknown feels should check the fields we know/care about and assert Eq correctly. the case that for comparing opaque unknown fields can be handled just as easily by comparing their byte repr directly, and is more clear to a user that you are not taking any semantic information into account by doing so (per your analogy, we are asserting that 42 is indeed 42, regardless of context). Doing Eq on a byte buffer containing a key and some bytes is equivalent to your implementation, minus validating that the key is valid.